### PR TITLE
PickList component changes

### PR DESCRIPTION
**Pull request by swethavemuri-anthem**
Converting Picklist component into a complex type to include @PicklistSelected so that a map for all code-label pairs will be available for the chosen items.

### CODE REVIEW

**Comment by openanthem-admin** [comment]
You did it @swethavemuri-anthem!

Thank you for signing the Anthem Opensource Software CLA.

File: nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts [hunk 1]
```diff
@@ -97,40 +105,46 @@ export class OrderablePickList extends BaseElement implements OnInit, ControlVal
...
+        if(this.element.leafState != null) {
             this.targetList = this.element.leafState;
         } else {
             this.targetList = [];
         }
+        if (this.targetList) {
+            for(var targetItem of this.targetList) {
+                this.parent.values = this.parent.values.filter(value => value.code != targetItem);
+            }
+        }
```
Comment (lopezton-anthem) [inline nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts]: This code is duplicated. Move common code into method and invoke method here.

File: nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts [hunk 2]
```diff
@@ -143,24 +157,39 @@ export class OrderablePickList extends BaseElement implements OnInit, ControlVal
...
+        this.pageService.eventUpdate$.subscribe(event => {
+            if(event.path == this.parent.path) {
+                // TODO write if condition to check if values have changed
+                    // When the source values change, validate that any values present in the target
+                    // are removed from the source values (if they are present)
+                    if (this.targetList) {
+                        for(var targetItem of this.targetList) {
+                            this.parent.values = this.parent.values.filter(value => value.code != targetItem);
+                        }
+                    }
```
Comment (lopezton-anthem) [inline nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts]: This code is duplicated. Move common code into method and invoke method here.

File: nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts [hunk 3]
```diff
@@ -250,4 +303,22 @@ export class OrderablePickList extends BaseElement implements OnInit, ControlVal
...
+    * This particularly comes into picture when the component is rendered in edit mode with pre loaded values
+    */
+   public getDesc(itm : any) : string {
+        let displayVal: string;
+        const values = this.selectedvalues;
+        values.forEach(value => {
+            if (value && value.code === itm) {
+                displayVal = value.label;
+            }
+        });
```
Comment (lopezton-anthem) [inline nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts]: Consider using `values.find` here.

File: nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts [hunk 4]
```diff
@@ -169,10 +198,34 @@ export class OrderablePickList extends BaseElement implements OnInit, ControlVal
...
+                }
+            });
+        } else {
+            if (this.value) {
+                this.element.leafState = this.value;
+            }
+        }
+        //    this.controlValueChanged.emit(this.parent);
+        // update the source values based on target list on load + onMoveToSource
+        // this.parent.values = this.removeduplicates(event.items[0]);
```
Comment (lopezton-anthem) [inline nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts]: remove commented code

File: nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/defn/ViewConfig.java [hunk 5]
```diff
@@ -1470,7 +1470,22 @@ public static boolean isValidNumericFilter(FilterMode mode) {
...
+	 * 	
+	 *  &#64;Getter &#64;Setter &#64;Type(SomeClass.class)
+	 *	public static class PicklistType {
+	 *	&#64;Values(value=AllCategory.class)
+	 *	&#64;Path("category")
+	 *	&#64;PickListSelected(postEventOnChange=true)
+	 *	private String[] selected; 
+	 *	}
+	 *	
+	 * </pre>
```
Comment (lopezton-anthem) [inline nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/defn/ViewConfig.java]: 👍 🥇 👏 👍 🥇 👏 👍 🥇 👏 👍 🥇 👏 👍 🥇 👏 👍 🥇 👏 👍 🥇 👏 👍 🥇 👏 👍 🥇 👏 👍 🥇 👏

File: nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/defn/ViewConfig.java [hunk 6]
```diff
@@ -1486,16 +1501,41 @@ public static boolean isValidNumericFilter(FilterMode mode) {
...
 		String labelClass() default "anthem-label";
 
-		boolean postEventOnChange() default false;
-
 		boolean readOnly() default false;
 
 		String sourceHeader() default "SourceList";
 
 		String targetHeader() default "TargetList";
+
```
Comment (lopezton-anthem) [inline nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/defn/ViewConfig.java]: Add Javadoc for this field to describe its behavior.

File: nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/defn/ViewConfig.java [hunk 7]
```diff
@@ -1486,16 +1501,41 @@ public static boolean isValidNumericFilter(FilterMode mode) {
...
-		boolean postEventOnChange() default false;
-
 		boolean readOnly() default false;
 
 		String sourceHeader() default "SourceList";
 
 		String targetHeader() default "TargetList";
+		
 		boolean showSourceControls() default false;
+
```
Comment (lopezton-anthem) [inline nimbus-core/src/main/java/com/antheminc/oss/nimbus/domain/defn/ViewConfig.java]: Add Javadoc for this field to describe its behavior.

File: nimbus-ui/nimbusui/src/app/components/platform/form-builder.service.ts [hunk 8]
```diff
@@ -54,7 +55,18 @@ export class FormElementsService {
...
+            const picklistparam: Param = element.type.model.params.find( p => 
+              p.alias === ViewComponent.selectedPicklist.toString());
+            var leafState = this._getTypeSafeLeafState(picklistparam);
+            if (checks) {
+              group[picklistparam.config.code] = [{value: leafState, disabled: !picklistparam.enabled}, checks];
+            } else {
+              group[picklistparam.config.code] = [{value: leafState, disabled: !picklistparam.enabled}];
+           }
+          }  
           //create new formgroup and formcontrol to create checkboxes in form. this is for form binding. TODO validations binding
```
Comment (lopezton-anthem) [inline nimbus-ui/nimbusui/src/app/components/platform/form-builder.service.ts]: Is TODO still pending?  If not, please remove this comment.

File: nimbus-ui/nimbusui/src/app/components/platform/form-element.component.html [hunk 9]
```diff
@@ -25,5 +24,11 @@
     <ng-template ngFor let-msg [ngForOf]="getMessages()">
         <nm-message [messageContext]="msg.context"  [messageArray]="msg.messageArray" [styleClass]="msg.styleClass"></nm-message>
     </ng-template>
- 
+    <ng-template [ngIf]="element?.config?.uiStyles?.attributes?.alias === 'PickList'">
+        <ng-template ngFor let-selectededparam [ngForOf]="element?.type?.model?.params">
+            <ng-template [ngIf]= "selectededparam?.config?.uiStyles?.attributes?.alias === 'PickListSelected'">
```
Comment (lopezton-anthem) [inline nimbus-ui/nimbusui/src/app/components/platform/form-element.component.html]: Use `ViewComponent.selectedPicklist.toString()` here.

File: nimbus-ui/nimbusui/src/app/components/platform/form-group.component.ts [hunk 10]
```diff
@@ -50,10 +50,8 @@ export const CUSTOM_INPUT_CONTROL_VALUE_ACCESSOR: any = {
                 <ng-template [ngIf]="!element.type?.model?.params?.length || element.config?.type?.collection">
                     <nm-element id="{{id}}" [element]="element" [elementCss]="elementCss" [form]="form"></nm-element>
                 </ng-template>
-                <ng-template [ngIf]="element.type?.model?.params?.length && element.config?.uiStyles?.attributes?.alias!=viewComponent.buttongroup.toString() && !element?.config?.type?.collection">
-                    <fieldset class="subQuestion" [hidden]="!element?.visible">
-                        <nm-frm-grp [elements]="element.type?.model?.params" [form]="form.controls[element.config?.code]" [elementCss]="elementCss" [parentElement]="element"></nm-frm-grp>
-                    </fieldset>
+                <ng-template [ngIf]="element?.config?.uiStyles?.attributes?.alias === 'PickList'">
```
Comment (lopezton-anthem) [inline nimbus-ui/nimbusui/src/app/components/platform/form-group.component.ts]: Use `ViewComponent.picklist.toString()` here.

File: nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts [hunk 11]
```diff
@@ -46,36 +48,42 @@ export const CUSTOM_INPUT_CONTROL_VALUE_ACCESSOR: any = {
...
+                </ng-template>
+            </p-pickList>   
+            </fieldset>   
         </div>
    `
 })
 
 export class OrderablePickList extends BaseElement implements OnInit, ControlValueAccessor {
-
     @Input() element: Param;
```
Comment (lopezton-anthem) [inline nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts]: The field `element` is already available in `BaseElement`. Consider removing `element` from `OrderablePickList`.

File: nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts [hunk 12]
```diff
@@ -169,10 +198,34 @@ export class OrderablePickList extends BaseElement implements OnInit, ControlVal
...
+                if (state && state.code) {
+                    const code = state.code;
+                    this.element.leafState.splice(i, 1, code);
+                }
+            });
+        } else {
+            if (this.value) {
+                this.element.leafState = this.value;
+            }
+        }
```
Comment (lopezton-anthem) [inline nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts]: Consider moving this code block and documentation into a private method.

File: nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts [hunk 13]
```diff
@@ -250,4 +303,22 @@ export class OrderablePickList extends BaseElement implements OnInit, ControlVal
...
-}
+    /*
+    * The method maps the code to a label based on the super set of @Values 
+    * passed as @Input("selectedValues") to the component.
+    * This particularly comes into picture when the component is rendered in edit mode with pre loaded values
+    */
+   public getDesc(itm : any) : string {
+        let displayVal: string;
+        const values = this.selectedvalues;
+        values.forEach(value => {
```
Comment (lopezton-anthem) [inline nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts]: Consider using `values.find` here.

File: nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts (hunk 12 above)
Comment (swethavemuri-anthem) [reply nimbus-ui/nimbusui/src/app/components/platform/form/elements/picklist.component.ts]: Addressed

**Pull request merged**